pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go

```diff
@@ -0,0 +1,221 @@
+	return MakeReplicationStreamError(ID, status)
+}
+
+func (p *partitionedStreamClient) Plan(ctx context.Context, ID StreamID) (Topology, error) {
+	db, err := gosql.Open("postgres", p.remote.String())
+	defer db.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := db.Conn(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	row := conn.QueryRowContext(ctx, `SELECT crdb_internal.replication_stream_spec($1)`, ID)
+	if row.Err() != nil {
+		return nil, errors.Wrap(row.Err(), "Error in planning a replication stream")
+	}
+
+	var rawSpec []byte
+	if err = row.Scan(&rawSpec); err != nil {
+		return nil, err
+	}
+	var spec streampb.ReplicationStreamSpec
+	if err := protoutil.Unmarshal(rawSpec, &spec); err != nil {
+		return nil, err
+	}
+
+	// we need to use the disqlplanner to find all the planner nodes and
+	// assign partition to it in a round-robin fashion
+	topology := Topology{}
+	for _, p := range spec.Partitions {
+		rawSpec, err := protoutil.Marshal(p.PartitionSpec)
+		if err != nil {
+			return nil, err
+		}
+		topology = append(topology, PartitionInfo{
+			ID:                p.NodeID.String(), // how do we determine partition id?
+			SubscriptionToken: SubscriptionToken(rawSpec),
+			SrcInstanceID:     int(p.NodeID),
+			SrcAddr:           streamingccl.PartitionAddress(p.SQLAddress.String()),
+			SrcLocality:       p.Locality,
+		})
+	}
+	return topology, nil
+}
+
+type partitionStreamDecoder struct {
+	rows *gosql.Rows
+	e    streampb.StreamEvent
+}
+
+func (d *partitionStreamDecoder) pop() streamingccl.Event {
+	if d.e.Checkpoint != nil {
+		event := streamingccl.MakeCheckpointEvent(d.e.Checkpoint.Spans[0].Timestamp)
+		d.e.Checkpoint = nil
+		return event
+	}
+
+	if d.e.Batch != nil {
+		event := streamingccl.MakeKVEvent(d.e.Batch.KeyValues[0])
+		d.e.Batch.KeyValues = d.e.Batch.KeyValues[1:]
+		if len(d.e.Batch.KeyValues) == 0 {
+			d.e.Batch = nil
+		}
+		return event
+	}
+
+	return nil
+}
+
+func (d *partitionStreamDecoder) decode() error {
+	var data []byte
+	if err := d.rows.Scan(&data); err != nil {
+		return err
+	}
+	var streamEvent streampb.StreamEvent
+	if streamEvent.Checkpoint == nil && streamEvent.Batch == nil {
+		return errors.New("unexpected event type")
+	}
+	d.e = streamEvent
+	return nil
+}
+
+func (p *partitionedStreamClient) Subscribe(
+	ctx context.Context, stream StreamID, spec SubscriptionToken, checkpoint hlc.Timestamp,
+) (chan streamingccl.Event, chan error, error) {
+
+	db, err := gosql.Open("postgres", p.remote.String())
+	if err != nil {
+		return nil, nil, err
+	}
+	conn, err := db.Conn(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sps := streampb.StreamPartitionSpec{}
+	if err = protoutil.Unmarshal(spec, &sps); err != nil {
+		return nil, nil, err
+	}
+	sps.StartFrom = checkpoint
+
+	specBytes, err := protoutil.Marshal(&sps)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rows, err := conn.QueryContext(ctx, `SELECT * FROM crdb_internal.stream_partition($1)`, specBytes)
```

Comment:
You must execute `SET avoid_buffering = true` before you can use stream_partition.

Reply:
(and you have to be rebased on top of my PR for this to work.